File: src/magma/api/Result.java
package magma.api;

import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

public interface Result<T, E> {
    Optional<T> findValue();

    <R> Result<R, E> mapValue(Function<T, R> mapper);

    <R> Result<T, R> mapErr(Function<E, R> mapper);

    <R> Result<Tuple<T, R>, E> and(Supplier<Result<R, E>> mapper);

    <R> Result<R, E> flatMapValue(Function<T, Result<R, E>> mapper);

    boolean isOk();

    boolean isErr();

    Optional<E> findError();

    <R> R match(Function<T, R> onOk, Function<E, R> onErr);

    T $() throws UnsafeException;

    <R> Result<T, R> replaceErr(Supplier<R> supplier);
}

File: src/clang/magma/result/Result.c
#include "Result.h"
magma.option.Option<T> findValue();
magma.option.Option<X> findError();
magma.result.Result<magma.result.R, X> mapValue(magma.result.R(*mapper)(T));
magma.result.Result<magma.result.R, X> flatMapValue(magma.result.Result<magma.result.R, X>(*mapper)(T));
magma.result.Result<T, magma.result.R> mapErr(magma.result.R(*mapper)(X));
magma.result.R match(magma.result.R(*whenOk)(T), magma.result.R(*whenErr)(X));
magma.result.Result<magma.option.Tuple<T, magma.result.R>, X> and(magma.result.Result<magma.result.R, X>(*other)());
magma.result.void consume(void(*whenOk)(T), void(*whenErr)(X));

